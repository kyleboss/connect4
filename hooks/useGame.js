import { useContext } from "react";

import GameContext from "@/state/GameContext";

const useGame = () => useContext(GameContext);

export default useGame;
