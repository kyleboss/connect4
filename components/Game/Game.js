import { useMemo } from "react";

import useGame from "@/hooks/useGame";
import Board from "@/components/Board";
import YourTurn from "@/components/YourTurn";
import WaitingForPlayers from "@/components/WaitingForPlayers";
import GameOver from "@/components/GameOver";
import NameTag from "@/components/NameTag";
import Button from "@/components/Button";
import {
  GAME_STATE_GAME_OVER,
  GAME_STATE_WAITING,
  USER_ROLE_GUEST,
} from "@/utils/constants";

const Game = () => {
  const { userRole, gameState, resetGame } = useGame();

  const isPlayer = useMemo(() => userRole !== USER_ROLE_GUEST, [userRole]);

  if (gameState === GAME_STATE_WAITING) {
    return <WaitingForPlayers />;
  }

  if (gameState === GAME_STATE_GAME_OVER) {
    return <GameOver />;
  }

  return (
    <>
      <NameTag />
      <YourTurn />
      <Board />
      {isPlayer ? <Button onClick={resetGame}>Reset Game</Button> : null}
    </>
  );
};

export default Game;
