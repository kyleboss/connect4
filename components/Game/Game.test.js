import React from "react";
import { render } from "@testing-library/react";

import Game from "./Game";
import useGame from "../../hooks/useGame";

jest.mock("../../hooks/useGame", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("../Board", () => {
  const Board = () => <div>Mocked Board</div>;
  Board.displayName = "Board";
  return Board;
});

jest.mock("../YourTurn", () => {
  const YourTurn = () => <div>Mocked YourTurn</div>;
  YourTurn.displayName = "YourTurn";
  return YourTurn;
});

describe("Game", () => {
  it("renders WaitingForPlayers when gameState is waiting", () => {
    useGame.mockImplementation(() => ({
      userRole: "Red",
      gameState: "GAME_STATE_WAITING",
      resetGame: jest.fn(),
    }));

    const { asFragment } = render(<Game />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders GameOver when gameState is game over", () => {
    useGame.mockImplementation(() => ({
      userRole: "Red",
      gameState: "GAME_STATE_GAME_OVER",
      resetGame: jest.fn(),
    }));

    const { asFragment } = render(<Game />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the game board, YourTurn and Reset button for players", () => {
    useGame.mockImplementation(() => ({
      userRole: "Red",
      gameState: "some_other_state",
      resetGame: jest.fn(),
    }));

    const { asFragment, getByText } = render(<Game />);
    expect(getByText("Mocked Board")).toBeInTheDocument();
    expect(getByText("Mocked YourTurn")).toBeInTheDocument();
    expect(getByText("Reset Game")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not render Reset button for guests", () => {
    useGame.mockImplementation(() => ({
      userRole: "USER_ROLE_GUEST",
      gameState: "some_other_state",
      resetGame: jest.fn(),
    }));

    const { queryByText } = render(<Game />);
    expect(queryByText("Reset Game")).not.toBeInTheDocument();
  });
});
