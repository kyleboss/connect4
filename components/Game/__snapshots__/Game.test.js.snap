// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders GameOver when gameState is game over 1`] = `
<DocumentFragment>
  <h1
    class="sc-69bb2827-0 iUjoYw"
  >
    Game over.
  </h1>
  <h2
    class="sc-b8c48683-0 iZmNFj"
  >
    undefined won.
  </h2>
  <button
    class="sc-c2948911-0 fZBMRl"
  >
    New Game
  </button>
</DocumentFragment>
`;

exports[`Game renders WaitingForPlayers when gameState is waiting 1`] = `
<DocumentFragment>
  <h1
    class="sc-69bb2827-0 iUjoYw"
  >
    Waiting for opponent...
  </h1>
</DocumentFragment>
`;

exports[`Game renders the game board, YourTurn and Reset button for players 1`] = `
<DocumentFragment>
  <h1
    class="sc-69bb2827-0 iUjoYw"
  />
  <div>
    Mocked YourTurn
  </div>
  <div>
    Mocked Board
  </div>
  <button
    class="sc-c2948911-0 fZBMRl"
  >
    Reset Game
  </button>
</DocumentFragment>
`;
